instrumentation/instafasthttp/client_test.go

```diff
@@ -0,0 +1,717 @@
+
+	go func() {
+		if err := server.Serve(ln); err != nil {
+			t.Errorf("unexpected error: %v", err)
+		}
+	}()
+
+	ln.Close()
+
+	c := &fasthttp.Client{
+		Dial: func(addr string) (net.Conn, error) { return ln.Dial() },
+	}
+	ic := instafasthttp.GetInstrumentedClient(s, c)
+
+	r := &fasthttp.Request{}
+	r.Header.SetMethod(fasthttp.MethodGet)
+	r.Header.Set("Authorization", "Basic blah")
+	r.URI().SetPath("/hello")
+	r.URI().SetQueryString("q=term&key=s3cr3t")
+	r.URI().SetHost("example.com")
+
+	resp := fasthttp.AcquireResponse()
+	defer fasthttp.ReleaseResponse(resp)
+
+	// Make the request
+	err := ic.Do(ctx, r, resp)
+
+	assert.Error(t, err)
+
+	spans := recorder.GetQueuedSpans()
+	require.Len(t, spans, 2)
+
+	span, logSpan := spans[0], spans[1]
+	assert.Equal(t, 1, span.Ec)
+	assert.EqualValues(t, instana.ExitSpanKind, span.Kind)
+
+	require.IsType(t, instana.HTTPSpanData{}, span.Data)
+	data := span.Data.(instana.HTTPSpanData)
+
+	assert.Equal(t, instana.HTTPSpanTags{
+		Method: "GET",
+		URL:    "http://example.com/hello",
+		Params: "key=%3Credacted%3E&q=term",
+		Error:  "InmemoryListener is already closed: use of closed network connection",
+	}, data.Tags)
+
+	assert.Equal(t, span.TraceID, logSpan.TraceID)
+	assert.Equal(t, span.SpanID, logSpan.ParentID)
+	assert.Equal(t, "log.go", logSpan.Name)
+
+	// assert that log message has been recorded within the span interval
+	assert.GreaterOrEqual(t, logSpan.Timestamp, span.Timestamp)
+	assert.LessOrEqual(t, logSpan.Duration, span.Duration)
+
+	require.IsType(t, instana.LogSpanData{}, logSpan.Data)
+	logData := logSpan.Data.(instana.LogSpanData)
+
+	assert.Equal(t, instana.LogSpanTags{
+		Level:   "ERROR",
+		Message: `error.object: "InmemoryListener is already closed: use of closed network connection"`,
+	}, logData.Tags)
+}
+
+func TestClient_DoTimeout(t *testing.T) {
+	recorder := instana.NewTestRecorder()
+	opts := &instana.Options{
+		Service: "test-service",
+		Tracer: instana.TracerOptions{
+			CollectableHTTPHeaders: []string{"x-custom-header-1", "x-custom-header-2"},
+		},
+		AgentClient: alwaysReadyClient{},
+	}
+	tracer := instana.NewTracerWithEverything(opts, recorder)
+	s := instana.NewSensorWithTracer(tracer)
+
+	parentSpan := tracer.StartSpan("parent")
+	ctx := instana.ContextWithSpan(context.Background(), parentSpan)
+
+	var fieldTFrmHeader, fieldSFrmHeader string
+
+	server := &fasthttp.Server{
+		Handler: func(ctx *fasthttp.RequestCtx) {
+			// get the header span and trace id from request header
+			fieldTFrmHeader = string(ctx.Request.Header.Peek(instana.FieldT))
+			fieldSFrmHeader = string(ctx.Request.Header.Peek(instana.FieldS))
+			ctx.Response.Header.Add("X-Response", "true")
+			ctx.Response.Header.Add("X-Custom-Header-2", "response")
+			ctx.Success("aaa/bbb", []byte("Ok response!"))
+		},
+	}
+
+	ln := fasthttputil.NewInmemoryListener()
+
+	go func() {
+		if err := server.Serve(ln); err != nil {
+			t.Errorf("unexpected error: %v", err)
+		}
+	}()
+
```

Comment:
I prefer to run the server setup within each test. This approach allows us to easily customise the server handlers for individual tests and also use separate, customised net.Listener instances for each server creation. It makes the server creation process more straightforward and simpler.

If we handle this in TestMain, the server creation itself would involve additional logic and we would also need to manage the listeners separately.

Moreover, as we are using an InMemoryListener instead of listening on an actual port on the machine, this approach is significantly faster and consumes minimal (or no) resources.